neat_ml/workflow/lib_workflow.py

```diff
@@ -0,0 +1,105 @@
+import logging
+from pathlib import Path
+from typing import Any, Optional
+import pandas as pd
+
+from neat_ml.opencv.preprocessing import process_directory as cv_preprocess
+from neat_ml.opencv.detection import collect_tiff_paths, run_opencv
+
+__all__ = ["get_path_structure", "stage_opencv", "stage_detect"]
+
+log = logging.getLogger(__name__)
+
+def get_path_structure(
+    roots: dict[str, str],
+    dataset_config: dict[str, Any],
+) -> dict[str, Path]:
+    """
+    Build only the paths needed by active steps.
+
+    Parameters
+    ----------
+    roots : dict[str, str]
+        Root dirs (work).
+    dataset_config : dict[str, Any]
+        Dataset dict (id, method, class, time_label, detection).
+
+    Returns
+    -------
+    paths : dict[str, Path]
+        Paths keyed by step usage (proc_dir, det_dir).
+    """
+    paths: dict[str, Path] = {}
+
+    ds_id: str = str(dataset_config.get("id", "unknown"))
+    method: str = str(dataset_config.get("method", ""))
+    class_label: str = str(dataset_config.get("class", ""))
+    time_label: str = str(dataset_config.get("time_label", ""))
+    work_root: Path = Path(roots["work"])
+
+    base_proc: Path = work_root / ds_id / method / class_label / time_label
+    paths["proc_dir"] = base_proc / f"{time_label}_Processed_{method}"
+    paths["det_dir"] = base_proc / f"{time_label}_Processed_{method}_With_Blob_Data"
+
+    return paths
+
+def stage_opencv(dataset_config: dict[str, Any], paths: dict[str, Path]) -> None:
+    """
+    Run OpenCV preprocessing + detection when configured.
+
+    Parameters
+    ----------
+    dataset_config : dict[str, Any]
+        Dataset config. Expects 'method' == 'OpenCV' and 'detection' block.
+    paths : dict[str, Path]
+        Paths from get_path_structure() (proc_dir, det_dir if built).
+    """
+    detection_cfg: dict[str, Any] = dict(dataset_config.get("detection", {}))
+    img_dir_str: Optional[str] = detection_cfg.get("img_dir")
```

Comment:
we don't need type hints in function bodies, just their signatures -- I won't leave any more comments about that, but please cleanup throughout the diff. You also have spurious coercions like the `dict()` one above that are not needed -- I'll let you check your diff a few more times and clean those out.